cdap-tms/src/main/java/co/cask/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java

```diff
@@ -230,7 +231,7 @@ private long fetchAndProcessMessages() throws TopicNotFoundException, IOExceptio
     // Process the notifications and record the message id of where the processing is up to.
     // 90% of the tx timeout is .9 * 1000 * txTimeoutSeconds = 900 * txTimeoutSeconds
     long timeBoundMillis = 900L * txTimeoutSeconds;
-    MessageTrackingIterator iterator = Transactionals.execute(getTransactional(), context -> {
+    MessageTrackingIterator iterator = Transactionals.execute(getTransactional(), txTimeoutSeconds, context -> {
```

Comment:
I think we should keep the logic restricted to this class. Introduce a new txTimeout variable that is not final. If we catch a transaction timeout exception in this loop, double that variable (up to some max, maybe 5 minutes). Once we process a batch of messages successfully, reset that txTimeout to the base tx timeout.